db/db_impl.cc

```diff
@@ -1639,6 +1641,271 @@ std::vector<Status> DBImpl::MultiGet(
   return stat_list;
 }
 
+// Order keys by CF ID, followed by key contents
+inline bool CompareKeyContext(const KeyContext* lhs, const KeyContext* rhs) {
+  auto cfh = reinterpret_cast<ColumnFamilyHandleImpl*>(lhs->column_family);
+  auto cfd = cfh->cfd();
+  auto cf_id1 = cfd->GetID();
+  cfh = reinterpret_cast<ColumnFamilyHandleImpl*>(rhs->column_family);
+  cfd = cfh->cfd();
+  auto cf_id2 = cfd->GetID();
+  if (cf_id1 < cf_id2) {
+    return true;
+  } else if (cf_id1 > cf_id2) {
+    return false;
+  }
+
+  auto comparator = cfd->user_comparator();
+  int cmp = comparator->Compare(*(lhs->key), *(rhs->key));
+  if (cmp < 0) {
+    return true;
+  }
+  return false;
+}
+
+void DBImpl::MultiGet(const ReadOptions& read_options,
+                      const std::vector<ColumnFamilyHandle*>& column_family,
+                      const std::vector<Slice>& keys, PinnableSlice* values,
+                      Status* statuses) {
+  autovector<KeyContext, MultiGetContext::MAX_KEYS_ON_STACK> key_context;
+  size_t index = 0;
+  for (const Slice& user_key : keys) {
+    key_context.emplace_back(user_key, column_family[index], &values[index],
+                             &statuses[index]);
+    index++;
+  }
+
+  MultiGetImpl(read_options, key_context);
+}
+
+void DBImpl::MultiGetImpl(const ReadOptions& read_options,
+      autovector<KeyContext, MultiGetContext::MAX_KEYS_ON_STACK>& key_context,
+      ReadCallback* callback, bool* is_blob_index) {
```

Comment:
Do we have two separate implementations for now? If yes, maybe it's a good idea to add a TODO comment mentioning the plan to consolidate them.

Reply:
Yes, the implementations are separate for now. I plan to eventually merge them after this is stabilized. One issue right now is there is some regression for small batches (i.e < 8) with no locality

Reply:
If there are two separate implementation, my suggest is to only support single column family MultiGet() in this new implementation. This will simplify a lot of things. We can even argue whether we want to implement a multi-CF MultiGet() at all. At least, MyRocks won't need it if it doesn't improve performance.

Reply:
Without Multi-CF, users cannot do a consistent read across column families, atleast not without an explicit snapshot. I'm not sure how important it is. Also, users will have to sort the keys by column family before calling MultiGet(), which will be a slight inconvenience.

Do you think either of these is an issue? If not, I agree we can drop multi-CF support in the new MultiGet and simplify things.